### PR TITLE
Fixing new iop_order for sidecar files.

### PR DESCRIPTION
**Pull request by jenshannoschwalm**
Discussing and debugging the new iop_order sheme showed there might be a problem
with the insertion of modules into the iop_order list.

Further tests here showed the new v3 algorithm is ok and gives correct iop_order
defined as a double.

But these tests also demonstrated, the exiv2 writing a float/double key to the sidecar uses low precision
1. resulting in iop_order of 38.9991 to be the same as 38.9992 **and** 
2. numbers close to a 'full' as 38.9999 to be in fact written as 39.0.

This means the priority of modules in sidecars is definitely wrong.
You **can** observe this by deleting the image from history and reimport it or just compare existing sidecar files with the data in main.history

This commit
- helps to debug and understand the problem, there is now
  #define DT_IOP_ORDER_INFO FALSE in iop_order.c
  If set to TRUE you will see what is inserted and what iop_order is given.
- Writes the iop_order **not** by toFloat but instead as a converted string with a precision of %.13f

Is this really enough? I don't really know. Possible problems:
- Already iop_order_v3 developed images deleted from the database will give wrong iop_orders
  when reimported.
- Do we also have to make sure exiv2 reads the high-precision iop_order as a double? Reading exiv2 docs about this did not give a definite answer for me and by experiment i would say it's ok as now.

### CODE REVIEW

**Comment by TurboGit** [comment]
I understand the issue but I'm not sure about the solution proposed here.

My first thinking on that was to re-compute the iop-order dynamically each time a new module is moved or added. In iop_order.c we do find the "right" place for a module in line 388. What about redistributing all iop-order in the current list ? This may also need to recomputing in the current dev history list? Not sure how easy this is or even if possible.

What do you think?

**Comment by jenshannoschwalm** [comment]
It's definitely **not** a **final** solution. it just is a temporary fix for problems we have atm as sidecar files have _vastly_ wrong information. When importing such data many things may happen.

About line 388, that's exactly the point so i added the DT_IOP_ORDER_INFO section right there :-)

So what is the problem of the algorithm?

When we use the new iop_order distribution algorithm we do lots of `_ioppr_insert_iop_before` and as the rules define this all takes place __very closely__. As we always divide by 2 one could say we populate a binary very densely at a small area. If we use a double float to represent the repeated divide we probably can safely do 40 insertions all together but as we don't keep track **where** we insert a number in that tree we do **not** know how many more we could do.

Anyway -- representing the iop_order as a float as done before when writing sidecars is definitely wrong. It just didn't become obvious so far as we didn't do so many `_ioppr_insert_iop_before`

Problems and how to solve
1. My proposal leaves room for some insertions even at the densest parts so it's a ok for a while.
2. Your idea about recalculating. Very tricky as we also have to respect data from xmp files and data available in main.history
3. The many iop_order checks do help as long as we write high precision data to xmp
4. We could introduce a counter for current code increasing with every iop_order insertion to make sure we don't overflow the binary divide.

File: src/common/exif.cc [hunk 1]
```diff
@@ -2918,7 +2918,14 @@ static void dt_set_xmp_dt_history(Exiv2::XmpData &xmpData, const int imgid, int
     snprintf(key, sizeof(key), "Xmp.darktable.history[%d]/darktable:multi_priority", num);
     xmpData[key] = multi_priority;
     snprintf(key, sizeof(key), "Xmp.darktable.history[%d]/darktable:iop_order", num);
-    xmpData[key] = iop_order;
+    // This code only writes iop_order as the default which is float, so it's wrong.
```
Comment (TurboGit) [inline src/common/exif.cc]: Ok, let's go with that but I would remove this comment which will be not clear when the code is removed. I would say:

// we ensure that we write the iop_order as high precision...

**Comment by jenshannoschwalm** [comment]
@TurboGit , i won't be at my pc until tomorrow so feel free to modify that yourself. otherwise i'll do it tomorrow.

There are several issues around that might be related. Unfortunately this pr does **not** fix the problem if you took the history data from an xmp file that has been written since introducing the new iop_order.
#3010 and #3009 are probably related.
When we have problems that could be related to history/iop_order we should always advice to discard history and try again after that to ensure the information in main.history is same as in xmp file.

I'll keep an eye open. There might be more coming. In all such issues we have to ask people for the data in main.history **and** xmp to understand a possible problem. We have the sidecar writing tool, could we extend that to also write the data from the history?

File: src/common/exif.cc (hunk 1 above)
Comment (jenshannoschwalm) [reply src/common/exif.cc]: Fixed.

**Comment by jenshannoschwalm** [comment]
good now.

**Comment by TurboGit** [comment]
Thanks ! Manually merged, the "..." at the end of my comment proposal was meant to be completed :)

**Comment by jenshannoschwalm** [comment]
@TurboGit and @aurelienpierre 

I suggest a surprisingly simple way to distribute the iop_orders more evenly without reordering.

I'm not doing a pr yet, just would like to hear what you both think because you seem to be the ones involved in this stuff.

See my branch relaxediop; the numbers shown for iop_order are **much** better.

Do i oversee any side effects? Would we need a new version?

**Pull request closed without merging**